lib/src/line_splitting/solve_state.dart

```diff
@@ -86,7 +83,7 @@ class SolveState {
 
   /// The constraints the bound rules in this state have on the remaining
   /// unbound rules.
-  Map<Rule, int> _constraints;
+  Map<Rule, int>? _constraints;
```

Comment:
I would recommend using a late final pattern here, e.g.:
```dart
late final Map<Rule, int> _constraints = _computeConstraints();
```

And then change _ensureConstraints to something like this:

```dart
Map<Rule, int> _computeConstraints() {
  _unboundRules = <Rule>{};
  _boundRules = <Rule>{};
  
  for (var rule in _splitter.rules) {
    ...
  }

  var constraints = <Rule, int>{};

  for (var bound in _boundRules) {
    ...
  }

  return constraints;
}
```

Then your use sites no longer have to call `_ensureConstraints`.  They can just refer to `_constraints` directly.

A similar comment applies to `_unboundConstraints` and `boundRulesInUnboundLines`.

Reply:
Love it.